Winleafs-Server/Winleafs.Server.Api/Controllers/SpotifyController.cs

```diff
@@ -0,0 +1,148 @@
+        private ISpotifyService _spotifyService;
+
+        public SpotifyController(DbContext context, IUserService userService, ISpotifyService spotifyService) : base(context)
+        {
+            _userService = userService;
+            _spotifyService = spotifyService;
+        }
+
+        /// <remarks>
+        /// Must be GET due to Spotify authorization standards
+        /// </remarks>
+        [HttpGet]
+        [Route("authorize/{applicationId}")]
+        public async Task<IActionResult> Authorize(string applicationId)
+        {
+            Guid guidOutput;
+            if (string.IsNullOrWhiteSpace(applicationId) || !Guid.TryParse(applicationId, out guidOutput))
+            {
+                return BadRequest("Error: Invalid parameters given."); //Generic error to not give too much information
+            }
+
+            //Add the user if it is his/her first time using the spotify functionality
+            await _userService.AddUserIfNotExists(applicationId);
+
+            const string scope = "playlist-read-private playlist-read-collaborative user-read-currently-playing user-read-playback-state";
+            return Redirect($"https://accounts.spotify.com/authorize/?client_id={SpotifyClientInfo.ClientID}&response_type=code&redirect_uri={SpotifyClientInfo.RedirectURI}&scope={scope}&state={applicationId}&show_dialog=false");
+        }
+
+        /// <remarks>
+        /// Must be GET due to Spotify authorization standards
+        /// </remarks>
+        [HttpGet]
+        [Route("swap")]
+        public async Task<IActionResult> SwapToken([FromQuery]string code, [FromQuery]string state, [FromQuery]string error)
+        {
+            //If there is an error, the user did not grant access
+            if (!string.IsNullOrEmpty(error))
+            {
+                return BadRequest("Error: Winleafs did not get access to Spotify.");
+            }
+
+            if (string.IsNullOrWhiteSpace(code) || string.IsNullOrWhiteSpace(state))
+            {
+                return BadRequest("Error: Invalid parameters given."); //Generic error to not give too much information
+            }
+
+            try
+            {
+                await _spotifyService.SwapToken(code, state); //We passed the application id as state for the authorize request
+            }
+            catch (InvalidApplicationIdException ex)
+            {
+                Log.Warning(ex, $"Failed to find a matching application id for application id {state}.");
+                return BadRequest("Failed to find a matching application id. Please disconnect from Spotify and try again.");
+            }
+            catch (Exception ex)
+            {
+                Log.Warning(ex, "Unknown error during Spotify authorization.");
+                return BadRequest("Unknown error during Spotify authorization. Please disconnect from Spotify and try again.");
+            }
+
+            return Ok("Authentication successful! You can now close this window");
```

Comment:
API window closing?

Reply:
What do you mean?

Reply:
"You can now close this window" but it comes as a response from the API?

Reply:
Ah, that is due to the way Spotify authorization works:
- User click authorize in the Winleafs app
- This opens a browser which does a get to /authorize
- This redirect to Spotify website, where the user can authenticate
- This redirects to /swap, which finalizes the authorization

But then a browser window is still open on the users side, hence we show this text to tell the user the authorization was successful and this browser tab/window can be safely closed.